### PR TITLE
NBCRESP-2411 - eliminate empty space at end of carousel

### PR DESCRIPTION
**Pull request by shane-gorilla**
Reimplements the necessities from https://github.com/nbcdotcom/react-slick/pull/8 and https://github.com/nbcdotcom/react-slick/pull/9

Takes inspiration from https://github.com/nbcdotcom/react-slick/pull/10 and hardens it when swiping the carousel.

Adds new `endRightEdge` prop to enable the behavior.

You can test this using the `npm start` local server

### CODE REVIEW

File: src/mixins/trackHelper.js [hunk 1]
```diff
@@ -152,5 +154,35 @@ export var getTrackLeft = function (spec) {
       }
   }
 
+  if (spec.endRightEdge === true && !spec.centerMode && !spec.infinite) {
```
Comment (acao) [inline src/mixins/trackHelper.js]: a one line comment might be nice to add here? just to explain whats going on?

File: src/mixins/trackHelper.js [hunk 2]
```diff
@@ -18,7 +18,10 @@ export var getTrackCSS = function(spec) {
   const trackChildren = (spec.slideCount + 2 * spec.slidesToShow);
 
   if (!spec.vertical) {
-    if (spec.variableWidth) {
+    if (spec.variableWidth && !spec.infinite) {
+      trackWidth = spec.slideCount * spec.slideWidth
```
Comment (acao) [inline src/mixins/trackHelper.js]: Good catch!

File: src/arrows.jsx [hunk 3]
```diff
@@ -39,10 +40,24 @@ export var PrevArrow = React.createClass({
...
 
 export var NextArrow = React.createClass({
+  getInitialState() {
+    return { disabled: false };
+  },
   clickHandler: function (options, e) {
     if (e) { e.preventDefault(); }
     this.props.clickHandler(options, e);
   },
+  componentWillReceiveProps() {
```
Comment (acao) [inline src/arrows.jsx]: nice! i was wondering how to do this.

File: src/mixins/helpers.js [hunk 4]
```diff
@@ -249,26 +254,66 @@ var helpers = {
...
 
     this.autoPlay();
   },
+  afterChange: function (previousSlide, currentSlide) {
+    var callback;
+
+    if (this.props.endRightEdge && !this.props.infinite && this.list && this.track) {
+      var {partiallyVisible, rightVisible, lastSlide, listRect} = getLastSlideVisibility(assign({listRef: this.list, trackRef: this.track}, this.props));
+      if (!partiallyVisible && rightVisible) {
+        var targetLeft = (lastSlide.offsetLeft - (listRect.width - lastSlide.clientWidth)) * -1;
```
Comment (acao) [inline src/mixins/helpers.js]: this algorithm makes another appearance! shall we export it?

File: src/mixins/trackHelper.js [hunk 5]
```diff
@@ -152,5 +154,35 @@ export var getTrackLeft = function (spec) {
...
+export var getMultiplier = function(spec) {
+  checkSpecKeys(spec, [
+    'children', 'currentSlide'
+  ]);
+
+  var slideComponent = spec.children[spec.currentSlide];
+  return slideComponent && slideComponent.props['data-speed-multiplier'] || 1;
+}
+
+export var getLastSlideVisibility = function (spec) {
```
Comment (acao) [inline src/mixins/trackHelper.js]: This function seems more and more popular. Are there any ways we can make it more efficient?

Comment (shane-gorilla) [inline src/mixins/trackHelper.js]: It's doing less work than most of the React Library calls. I think the performance here is acceptable. 
screen shot 2017-02-07 at 11 06 15 am

File: src/mixins/helpers.js (hunk 4 above)
Comment (shane-gorilla) [inline src/mixins/helpers.js]: Sure thing!

File: src/mixins/trackHelper.js (hunk 1 above)
Comment (shane-gorilla) [inline src/mixins/trackHelper.js]: Sure thing!

File: src/mixins/trackHelper.js (hunk 5 above)
Comment (shane-gorilla) [inline src/mixins/trackHelper.js]: Do you think there's a better way to go about this one?

Comment (acao) [inline src/mixins/trackHelper.js]: looks good!

**Pull request merged**